### PR TITLE
Use io.Pipe in PostForm to avoid reading whole files into memory

### PR DESCRIPTION
**Pull request by a-palchikov**
Implement PostFrom in terms of `io.Pipe`. This will be beneficial for uploading large(r) files.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -167,44 +167,69 @@ func (c *Client) PostForm(endpoint string, vals url.Values, files ...File) (*Res
...
+				return
 			}
-		}
-		boundary := writer.Boundary()
-		if err := writer.Close(); err != nil {
-			return nil, err
-		}
-		req, err := http.NewRequest("POST", endpoint, body)
+
+			errCh <- writer.Close()
```
Comment (klizhentas) [inline client.go]: the significance of this channel is not clear to me - why not use `writer.CloseWithError(err)` instead? The error should pop up on the other side of the wire:

https://golang.org/pkg/io/#PipeWriter.CloseWithError

File: client.go [hunk 2]
```diff
@@ -167,44 +167,69 @@ func (c *Client) PostForm(endpoint string, vals url.Values, files ...File) (*Res
...
+		}()
+
+		req, err := http.NewRequest("POST", endpoint, readBody)
 		if err != nil {
+			readBody.Close()
+			<-errCh
```
Comment (klizhentas) [inline client.go]: not sure why do we need to wait here for this channel to complete?

File: client.go [hunk 3]
```diff
@@ -167,44 +167,69 @@ func (c *Client) PostForm(endpoint string, vals url.Values, files ...File) (*Res
...
 			return nil, err
 		}
 		c.addAuth(req)
 		req.Header.Set("Content-Type",
-			fmt.Sprintf(`multipart/form-data;boundary="%v"`, boundary))
-		return c.client.Do(req)
+			fmt.Sprintf(`multipart/form-data;boundary="%v"`, writer.Boundary()))
+		resp, err := c.client.Do(req)
+		errSend := <-errCh
```
Comment (klizhentas) [inline client.go]: not sure why wait here, I think that `CloseWithError` on write should do the trick.

**Comment by klizhentas** [comment]
I would like to add one more thing to this PR before merging it:

* With your change upload form becomes optimized for large file uploads and it introduces significant overhead for smaller requests. 
* I would like to fix this by using a combination of `io.LimitReader` and your pipe approach.
* If the request form is smaller than let's say 64KB, use LimitReader to read it all in memory. 
* Otherwise, discard the in memory buffer and replay the form into the pipe read writer.

**Comment by a-palchikov** [comment]
Addressed the comments but at the expense of a considerable increase in complexity.

**Comment by a-palchikov** [comment]
I've reshaped the implementation somewhat but not drastically:
  - I need a way to limit the first attempt to write the URL values / files into memory - thus the `limitWriter` abstraction.
  - I need a way to cache and replay data from multiple `File` objects - thus the `fileBuffer` abstraction.
  - I tried making the separation between initial write attempt and switch to a pipe more obvious by splitting  latter into a separate method.

File: client.go [hunk 4]
```diff
@@ -382,6 +379,32 @@ func (c *Client) addAuth(r *http.Request) {
 	}
 }
 
+func (c *Client) writeWithPipe(endpoint string, vals url.Values, buffers ...*fileBuffer) (*http.Response, error) {
+	r, w := io.Pipe()
+	writer := multipart.NewWriter(w)
+
+	for _, buffer := range buffers {
+		// Combine cache and the rest of the file
+		buffer.Reader = io.MultiReader(buffer.cache, buffer.f.Reader)
```
Comment (klizhentas) [inline client.go]: I expect this logic to be encapsulated in fileBuffer struct and not here

File: client.go [hunk 5]
```diff
@@ -479,3 +502,80 @@ type bearerAuth struct {
...
+		err = errWrite
+	}
+	return n, err
+}
+
+// limitWriter reads upto n bytes from the specified stream and returns errShortWrite
+// if more than n bytes are written
+type limitWriter struct {
+	w io.Writer
+	n int64
```
Comment (klizhentas) [inline client.go]: n -> to something more reasonable, e.g. maxBytes

File: client.go [hunk 6]
```diff
@@ -479,3 +502,80 @@ type bearerAuth struct {
...
+	if errWrite != nil {
+		err = errWrite
+	}
+	return n, err
+}
+
+// limitWriter reads upto n bytes from the specified stream and returns errShortWrite
+// if more than n bytes are written
+type limitWriter struct {
+	w io.Writer
```
Comment (klizhentas) [inline client.go]: w -> to writer

File: client.go [hunk 7]
```diff
@@ -479,3 +502,80 @@ type bearerAuth struct {
...
+
+func (r *fileBuffer) create(w *multipart.Writer) (io.Writer, error) {
+	return w.CreateFormFile(r.f.Name, r.f.Filename)
+}
+
+// fileBuffer is a File wrapper that buffers data from the specified File
+type fileBuffer struct {
+	io.Reader
+	cache *bytes.Buffer
+	f     File
```
Comment (klizhentas) [inline client.go]: f to file

File: client.go [hunk 8]
```diff
@@ -167,40 +168,36 @@ func (c *Client) PostForm(endpoint string, vals url.Values, files ...File) (*Res
...
-			_, err = io.Copy(w, f.Reader)
-			if err != nil {
-				return nil, err
-			}
+		var buf bytes.Buffer
+		buf.Grow(bufferSize)
+
+		// Cache file reads in case we reach the memory limit
+		// and need to rewind
+		buffers := make([]*fileBuffer, 0, len(files))
```
Comment (klizhentas) [inline client.go]: why pointer to a buffer

**Pull request merged**